### PR TITLE
Updated developer documenation

### LINKED ISSUE #212: Add section to developer guide on testing operations on images, etc.
Often we will want to write a unit test for some function that operates on a large structure, such as an image.

A place has been provided to include test data (sunpy.data.test) for these kinds of operations, but so far now rules have been created about what should be included there.

It would therefore be helpful to decide on such rules and include a section in the developers guide on the topic.

Things to consider:
1. When should a new data file be added to sunpy.data.test?
2. What about hard-coding "fake data" (e.g. a 2-dimensional ndarray) instead of reading in a file each time? For example,
   we could include some sample 32x32 pixel image data, base64-encoded in sunpy.data.test.
3. Limits on filesize?
4. Optimal dimensions to ensure accurate testing? (16x16? 32x32?)

It's important to work this out early on so that developers can write good test code for data operations.

### LINKED ISSUE #856: Add a section to the developer documentation about documentation
How to API doc using automodapi and some of the gotchas of automodapi and numpydoc is probably a good idea.

### LINKED ISSUE #1618: Don't clone into $HOME/sunpy
The fact that cloning and installing sunpy in $HOME/sunpy is a bad idea because it conflicts with the default data dir needs documenting better, and potentially we could check for this in `config.py`.

### LINKED ISSUE #1858: find a way to not get astropy-helpers in commits everywhere
Should we add astropy_helpers to `.gititnore`? that way we will only add it when it's needed and we would avoid this problem once and again, or add `ignore=dirty` on the `.gitmodules`?
- [ignore changes in submodules](http://www.nils-haldenwang.de/frameworks-and-tools/git/how-to-ignore-changes-in-git-submodules)
- [git submodules trips](http://blogs.atlassian.com/2013/03/git-submodules-workflows-tips/#tip4)

### LINKED ISSUE #2925: Add section to developer docs on running figure tests
It's not obvious to me how to run all the figure tests; I guessed from https://docs.sunpy.org/en/latest/dev_guide/tests.html#running-unit-tests that maybe it was
```
py.test -k figure
```
but they're all skipped if I run this and I'm not sure how to get them to run.

### LINKED ISSUE #2931: We should link to the citation instructions from the docs sidebar
It shouldn't duplicate the page, but it should be in the main sidebar for increased visibility.

### PR DESCRIPTION
**Pull request by nabobalis**
### Description

This is part 1 of the epic journey that is, "Nabil goes over all the files in SunPy".

Merged in #2934 
Fixes #2931 
Fixes #2925 
Fixes #1858 
Fixes #856 
Fixes #1618 
Fixes #212

### CONTRIBUTING.rst is the one to check the most

The other changes are mainly within the `dev_guide` section that hopefully explain more about our test suite and how to write/run aspects of SunPy

[Built docs.](https://circleci.com/gh/nabobalis/sunpy/4422/artifacts/0//root/project/docs/_build/html/index.html)

### CODE REVIEW

File: tox.ini [hunk 1]
```diff
@@ -41,7 +41,7 @@ deps =
 commands =
     offline,astropydev,numpydev: pytest -m "not figure" --cov=./sunpy {posargs}
     online: pytest --timeout=500 -m "not figure" --remote-data=any --cov=./sunpy {posargs}
-    build_docs: sphinx-build docs docs/_build/html -W -b html
```
Comment (nabobalis) [inline tox.ini]: This is removed until I fix the stability page.

**Comment by nabobalis** [comment]
There are enough changes in this PR that I won't correct the issues I spot with the API docs until PR number 2.

**Comment by nabobalis** [comment]
https://circleci.com/gh/nabobalis/sunpy/3539/artifacts/0//root/project/docs/_build/html/index.html

File: docs/dev_guide/code_standards.rst [hunk 2]
```diff
@@ -1,49 +1,107 @@
...
-
-Additionally, all code that goes in the project should be checked using one of the many PEP8 linters that are available (pep8, pylint, flake8 as some examples).
-This is to ensure that the SunPy code follows The Style Guide for Python (`PEP 8 <https://www.python.org/dev/peps/pep-0008/>`_).
-
-To ensure this, each pull request is checked for PEP8 issues by a bot called `pep8speaks <https://pep8speaks.com/>`_.
-This way, people are aware of any potential issues with their submitted code.
+=================
+
+* All code must be compatible with Python 3.6 and later.
+  Usage of ``six``, ``__future__``, and ``2to3`` is not longer acceptable.
```
Comment (wafels) [inline docs/dev_guide/code_standards.rst]: Usage of ``six``, ``__future__``, and ``2to3`` is no longer acceptable.

File: CONTRIBUTING.rst [hunk 3]
```diff
@@ -1,87 +1,246 @@
...
+Send it back to us
+^^^^^^^^^^^^^^^^^^
+
+Once you have some changes you would like to submit, you will need to commit the changes.
+This is a three stage process:
+
+1. Use ``git status`` to see that the only changes locally are the right ones.
+2. Use ``git add <path to file>`` to add the changes to `git`.
+3. Use ``git commit -m <message>`` to label those changes.
+
```
[the remark this answers is not in the document]
Comment (nabobalis) [reply CONTRIBUTING.rst]: Yes.

File: CONTRIBUTING.rst [hunk 4]
```diff
@@ -1,87 +1,246 @@
...
-For a good example of how to do this see issue `#1191`_.
+When reporting an issue, please try to provide a short description of the issue with a small code sample, this way we can attempt to reproduce the error.
+Also provide any error output generated when you encountered the issue, we can use this information to debug the issue.
+For a good example of how to do this see issue `#2879`_.
 
 If you are making a feature request, please post as much information as possible regarding the feature you would like to see in SunPy.
 
-Finding Things to Contribute to
--------------------------------
+We provide GitHub templates for both of these types of issue.
```
[the remark this answers is not in the document]
Comment (nabobalis) [reply CONTRIBUTING.rst]: I hopefully have clarified this section.

File: CONTRIBUTING.rst [hunk 5]
```diff
@@ -1,87 +1,246 @@
...
 
-Issue reports should provide a short description of the issue.
-If you are reporting a bug, please post a code sample so others can attempt to reproduce the error.
-Please also post any error output generated by the bug.
-For a good example of how to do this see issue `#1191`_.
+When reporting an issue, please try to provide a short description of the issue with a small code sample, this way we can attempt to reproduce the error.
+Also provide any error output generated when you encountered the issue, we can use this information to debug the issue.
+For a good example of how to do this see issue `#2879`_.
 
 If you are making a feature request, please post as much information as possible regarding the feature you would like to see in SunPy.
```
[the remark this answers is not in the document]
Comment (nabobalis) [reply CONTRIBUTING.rst]: Done.

File: CONTRIBUTING.rst [hunk 6]
```diff
@@ -1,87 +1,246 @@
...
+Do one final check to make sure the code changes look correct and then press the green "Create pull request" button.
+
+When you open your pull request, we have a GitHub template that will guide you on what to write in the message box.
+Please fill this in and title the pull request.
+Now the final step is to press the green "Create pull request" button.
+
+As soon as you do this, you will be greeted by a message from the "sunpy bot" as well as several continuous integration checks.
+These are explained on our :ref: `Version Control <version_control>` page.
+But what is important to know is that these run a series of tests to make sure that the changes do not cause any new errors.
+Now we (the SunPy community) can review the code and offer suggestions and once we are happy, we can merge in the pull request.
```
[the remark this answers is not in the document]
Comment (nabobalis) [reply CONTRIBUTING.rst]: I've expanded upon this section.

[the remark this answers is not in the document]
Comment (nabobalis) [reply CONTRIBUTING.rst]: I am actually wondering if I need to add a new subsection on how to run the tests

File: docs/dev_guide/code_standards.rst (hunk 2 above)
Comment (nabobalis) [reply docs/dev_guide/code_standards.rst]: Fixed.

File: CONTRIBUTING.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (nabobalis) [reply CONTRIBUTING.rst]: I decided to add it.

File: docs/dev_guide/maintainer_workflow.rst [hunk 7]
```diff
@@ -6,7 +6,7 @@ Workflow for Maintainers
 
 This page is for maintainers who can merge our own or other peoples' changes into the upstream repository.
 
-Being as how you're a maintainer, you are completely on top of the basic git stuff in :ref:`newcomers`.
+Being as how you're a maintainer, you are completely on top of the basic git stuff in :ref:`newcomers` and `Astropy's git workflow <http://docs.astropy.org/en/stable/development/workflow/development_workflow.html#development-workflow>`_.
```
Comment (Cadair) [inline docs/dev_guide/maintainer_workflow.rst]: Seeing as

**Comment by nabobalis** [comment]
I'm gonna merge this in now and if any problems arise, we can fix them after.

**Pull request merged**